File: src/scripts/ingest.ts
import { PDFLoader } from "langchain/document_loaders/fs/pdf";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { VectorStoreService } from "../services/VectorStoreService";
import path from "path";
import * as dotenv from "dotenv";

dotenv.config();

export async function ingestDocs() {
    console.log("Starting ingestion...");

    const pdfPath = path.join(__dirname, "../../kcg.pdf");
    console.log(`Loading PDF from: ${pdfPath}`);

    try {
        const loader = new PDFLoader(pdfPath);
        const docs = await loader.load();
        console.log(`Loaded ${docs.length} pages.`);

        const splitter = new RecursiveCharacterTextSplitter({
            chunkSize: 1000,
            chunkOverlap: 200,
        });

        const splitDocs = await splitter.splitDocuments(docs);
        console.log(`Split into ${splitDocs.length} chunks.`);

        const vectorStoreService = new VectorStoreService();
        await vectorStoreService.addDocuments(splitDocs);

        console.log("Ingestion complete!");
    } catch (error) {
        console.error("Ingestion failed:", error);
    }
}

File: fastapi_backend/scripts/ingest.py
import os
import asyncio
from langchain_community.document_loaders import PyPDFLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from ..services.vector_store_service import VectorStoreService
from dotenv import load_dotenv

# Load environment variables.
load_dotenv()

# The ingest_docs function processes our PDF file and puts it into the 'search engine'.
# For a beginner: This is like reading a whole book, cutting it into small 
# useful pieces, and indexing them so they are easy to find later.
async def ingest_docs():
    print("Starting ingestion...")

    # Define the path to our knowledge PDF.
    pdf_path = os.path.join(os.path.dirname(__file__), "../../kcg.pdf")
    print(f"Loading PDF from: {pdf_path}")

    try:
        # 1. LOAD: Read the PDF file.
        loader = PyPDFLoader(pdf_path)
        pages = await asyncio.to_thread(loader.load) # Run loading in a thread to keep it fast.
        print(f"Loaded {len(pages)} pages.")

        # 2. SPLIT: Break the long pages into smaller 'chunks'.
        # For a beginner: AI works best with small, focused pieces of text 
        # (about 1000 characters each) rather than whole pages.
        splitter = RecursiveCharacterTextSplitter(
            chunk_size=1000,
            chunk_overlap=200, # Overlap ensures no information is cut off at the edges.
        )
        split_docs = splitter.split_documents(pages)
        print(f"Split into {len(split_docs)} chunks.")

        # 3. STORE: Save these chunks into our ChromaDB vector database.
        vector_store_service = VectorStoreService()
        print(split_docs,'split docs')
        await vector_store_service.add_documents(split_docs)

        print("Ingestion complete! Your bot is now smarter.")
        
    except Exception as error:
        print(f"Ingestion failed: {error}")

# This part allows you to run the script directly from the terminal.
if __name__ == "__main__":
    # To run: python -m fastapi_backend.scripts.ingest
    asyncio.run(ingest_docs())
